src/marqo/tensor_search/throttling/redis_throttle.py

```diff
@@ -0,0 +1,110 @@
+                return function(*args, **kwargs)
+
+            redis = redis_driver.get_db()  # redis instance
+
+            logger.info(f"Beginning throttling check. API request is of type {request_type}")
+            lua_shas = redis_driver.get_lua_shas()
+
+            # Define maximum thread counts
+            throttling_max_threads = {
+                RequestType.INDEX: utils.read_env_vars_and_defaults(EnvVars.MARQO_MAX_CONCURRENT_INDEX),
+                RequestType.SEARCH: utils.read_env_vars_and_defaults(EnvVars.MARQO_MAX_CONCURRENT_SEARCH) 
+            }
+            
+            set_key = f"set:{request_type}"
+            thread_name = f"thread:{uuid.uuid4()}"
+
+            t0 = time.time()
+
+            # Check current thread count / increment using LUA script
+            try:
+                check_result = redis.evalsha(
+                    lua_shas["check_and_increment"], 
+                    1,          
+                    set_key,                                 # sorted set key (by request type)
+                    thread_name,                             # name of member for the thread
+                    throttling_max_threads[request_type],    # thread_limit
+                    utils.read_env_vars_and_defaults(EnvVars.MARQO_THREAD_EXPIRY_TIME)  # expire_time
+                )
+            except Exception as e:
+                logger.warn(f"There is a problem with your redis instance. Skipping throttling check. Reason: {e}")
+                redis_driver.set_faulty(True)
+                return function(*args, **kwargs)
+
+            t1 = time.time()
+            redis_time = (t1 - t0)*1000
+            
+            """
+            Only for testing
+            logger.info(f"Redis check and increment took {redis_time}ms.")
+            check_test_data = {
+                "timestamp": time.asctime(),
+                "action": "check",
+                "thread_name": thread_name,
+                "result": result_word,
+                "max_threads": throttling_max_threads[request_type],
+                "redis_time": redis_time
+            }
+            logger.info(f"DEBUG: redis message {check_result}")
+            """
+
+            # Thread limit exceeded, throw 429
+            if check_result == 0:
+                throttling_message = f"Throttled because maximum thread count ({throttling_max_threads[request_type]}) for request type '{request_type}' has been exceeded. Try your request again later."
+                raise TooManyRequestsError(message=throttling_message)
+
+            else:
+                # Execute function
+                try:
+                    result = function(*args, **kwargs)
+                    return result
+
+                except Exception as e:
+                    raise e
+                
+                # Delete thread key whether function succeeds or fails (async)
+                finally:
+
+                    def remove_thread_from_set(key, name):

```

Comment:
nit: consider moving this to the outer scope. Too many nested functions makes testing harder